### ISSUE TITLE
Access to websocket server materialized value

### ISSUE THREAD

**Issue by mathbruyen** [akka/akka-http#1167]
Websocket server methods do not allow accessing the materialized value of the flow it is given. They take a flow as argument and return the HTTP response to send back:

```java
HttpResponse response = WebSocket.handleWebSocketRequestWith(request, greeterFlow);
```

The server takes care of materializing the flow and ignores both values. This prevents users from getting them, which can be problematic. For example if the flow source is a queue, one needs to get its materialized value in order to push to the queue. A workaround is to setup a callback on `mapMaterializedValue` to send the queue to whoever is interested in it:

```java
Source.<Message>queue(5, OverflowStrategy.backpressure())
    .mapMaterializedValue(sourceQueue -> { /* send queue to whoever is interested */; return sourceQueue; });
```

But this is a workaround rather than a clean solution, as `mapMaterializedValue` is meant to transform the materialized value rather than broadcast it. The client has methods that return this materialized value:

```java
Pair<CompletionStage<WebSocketUpgradeResponse>, SourceQueueWithComplete<Message>> pair =
        http.singleWebSocketRequest(WebSocketRequest.create(connect), flow, materializer);
```

It would be great if the server had methods for doing something similar. This might be tricky since contrary to the client, the server takes care of materializing the value after the response it sent.

(mailing list pointer: [Graph materialized value of websocket upgrade](https://groups.google.com/forum/#!topic/akka-user/p4HDrgmu6fo))

**Comment by jrudolph** [comment]
I agree, @mathbruyen, that it would be nice to get access to the materialized value. Right now the recommended way is indeed using `mapMaterializedValue`. Do you have an idea how the API would look like? E.g. we could add another method `handleWebSocketRequestWithMat` that also returns a `Future[Mat]`. Would that work for you?

**Comment by ktoso** [comment]
Yeah, I think a method like `handleWebSocketRequestWithMat: (HttpResponse, Future[Mat])` would solve this and be pretty consistent with other APIs we have

**Comment by mathbruyen** [comment]
It would work for me yes, supposedly returning `Pair[HttpResponse, Future[Mat]]` in order to make response to send available as well.

**Comment by sportanova** [comment]
Would ```handleWebSocketRequestWithMat``` have to do something like [singleWebSocketRequest](https://github.com/akka/akka-http/blob/2e0d2b5b219c21ce3acd1ced0cb2a2e7cd6aa1ee/akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala#L608) ```.joinMat(clientFlow)(Keep.both).run()``` to get a ```(_, Mat)``` and then wrap the ```Mat``` in a future?

I'm new to akka-http but this looks fun and I'd like to take a crack it it

**Comment by sportanova** [comment]
or ```handleMessages``` needs to be changed to include the ```Future[Mat] ``` https://github.com/akka/akka-http/blob/5932237a86a432d623fafb1e84eeeff56d7485fe/akka-http-core/src/main/scala/akka/http/scaladsl/model/ws/UpgradeToWebSocket.scala#L37

**Comment by jrudolph** [comment]
Yes, there are several layers of APIs that could be touched. I think it makes most sense to add those methods to `UpgradeToWebSocket`. It can be implemented there directly by piggy-backing on the `mapMaterializedValue` trick.

Basically like this:

```scala
def handleMessagesMat[Mat](
                      handlerFlow: Graph[FlowShape[Message, Message], Mat],
                      subprotocol: Option[String]                          = None): (HttpResponse, Future[Mat]) = {
    val promise = Promise[Mat](O)
    val response = handleMessages(Flow.fromGraph(handlerFlow).mapMaterializedValue(mat => promise.success(mat)), subprotocol)
    (response, promise.future)
  }
```
